SwiftCompilerSources/Sources/Optimizer/Utilities/WalkUtils.swift

```diff
@@ -0,0 +1,518 @@
+
+import SIL
+
+
+/// - A `DefUseWalker` finds all uses of a target value.
+///
+/// - A target value is described by a "base" value and a projection path.
+///   1. If the projection path is empty (`""`) then the target value is the base value itself.
+///   2. If the projection path is non-empty (`"s0.1.e3"`), then the target value is the one
+///     reachable from the base through the series of projections described by the path.
+/// - A path can also contain a pattern such as `"v**"` which means any series of "value"
+///   projections (excluding `ref_element_addr` and similar, i.e. `c*`) from any field.
+///   In the `v**` case, the target value*s* are many, i.e. all the ones reachable from
+///   the base through _any of the fields_ through _any number_ of value projections.
+///   `c*` means values reachable through a _single_ projection of _any_ of the fields of the class.
+///
+/// - A walk is started with a call to `walkDownUses(base, path: path, state: state)`.
+/// - This function will call `walkDown(operand, path: path, state: state)`
+///   for every use of `base` as `operand` in an instruction.
+/// - For each use, then the walk can continue with base the result if the result of the using
+///   instruction might still reach the target value with a new projection path.
+///   1. If the use is a construction such as a
+///     `%res = struct $S (%f0)` (or `%res = tuple (%unk, %1)`) instruction and the path is `p`
+///     then the `%res` result value reaches the target value through the new projection`s0.p` (respectively `1.p`).
+///   2.  If the use is a projection such as `%res = struct_extract %s : $S, #S.field0` and the
+///     path is `s0.s1` then the target value is reachable from `%res` with path `s1`.
+///     If the path doesn't match `unmatchedPath` is called.
+///   3. If the use is a "forwarding instruction", such as a cast, the walk continues with the same path.
+///   4. If the use is an unhandled instruction then `leafUse` is called to denote that the client has to
+///     handle this use.
+///
+/// - `State` can be defined by the implementor to track specific state of a "branch"
+///   of the walk, i.e. whether a certain instruction was crossed while walking towards _this use_ of the target.
+///
+/// There are two types of DefUseWalkers, one for values (`ValueDefUseWalker`) and one for
+/// addresses (`AddressDefUseWalker`)
+
+
+/// A `ValueDefUseWalker` can only handle "value" bases, which correspond
+/// to types that are not addresses, i.e. _do not have_ an asterisk (`*`) in the textual
+/// representation of their SIL type (`$T`).
+/// These can be values of reference type, or struct/tuple etc.
+/// A `ValueDefUseWalker.walkDownDefault` called on a use of a "value" base which
+/// yields an "address" value (such as `ref_element_addr %value_base`) will call `leafUse`
+/// since the walk can't proceed.
+/// **All functions return a boolean flag which, if true, can stop the walk of the other uses
```

Comment:
> The return value should indicate whether the action was succesfull.

I agree.

But it would be even better to return an enum instead of a Bool, e.g.
```
enum WalkResult {
  case continueWalk
  case abortWalk
}
```

Returning a Bool is fine for functions which imply the meaning by it's name, e.g. `isEscaping`.
But for functions like `walkDown`, it's better to return an enum, which makes the meaning of the returned value immediately clear.